### PR TITLE
Implement Job Template Sharding/Splitting/Slicing

### PR DESCRIPTION
**Pull request by matburt**
This is a work-in-progress, see https://github.com/ansible/awx/issues/1283

### CODE REVIEW

File: awx/main/models/workflow.py [hunk 1]
```diff
@@ -433,6 +433,14 @@ class Meta:
...
         on_delete=models.SET_NULL,
     )
+    job_template = models.ForeignKey(
+        'JobTemplate',
+        related_name='sharded_jobs',
+        blank=True,
+        null=True,
+        default=None,
+        on_delete=models.SET_NULL,
+    )
```
Comment (AlanCoding) [inline awx/main/models/workflow.py:466]: I would prefer a field name that include any information about its function, like `shard_job_template`.

File: awx/main/models/jobs.py [hunk 2]
```diff
@@ -277,6 +277,12 @@ class Meta:
         default=False,
         allows_field='credentials'
     )
+    job_shard_count = models.IntegerField(
+        blank=True,
+        default=0,
+        help_text=_("The number of jobs to split into at runtime. Will cause the Job Template to launch a workflow."),
```
Comment (AlanCoding) [inline awx/main/models/jobs.py]: best to note in help text that 0 indicates that no workflow will be spawned, which is different than if you made it 1.

File: awx/api/serializers.py [hunk 3]
```diff
@@ -3009,7 +3009,7 @@ class Meta:
         fields = ('*', 'host_config_key', 'ask_diff_mode_on_launch', 'ask_variables_on_launch', 'ask_limit_on_launch', 'ask_tags_on_launch',
                   'ask_skip_tags_on_launch', 'ask_job_type_on_launch', 'ask_verbosity_on_launch', 'ask_inventory_on_launch',
                   'ask_credential_on_launch', 'survey_enabled', 'become_enabled', 'diff_mode',
-                  'allow_simultaneous', 'custom_virtualenv')
+                  'allow_simultaneous', 'custom_virtualenv', 'job_shard_count')
```
Comment (AlanCoding) [inline awx/api/serializers.py]: In the JT serializer, we should considering throwing a 400 if `allow_simultaneous` is false and `job_shard_count` is greater than 1.

I can't envision a reasonable use case for this configuration of linearizing a JT, and I'd predict some user confusion if we allow it.

----

Note:

This was resolved by making a different design decision. The sharded jobs will _always_ have `allow_simultaneous=True`. The _workflow shard containers_ may or may not be simultaneous. This is actually consistent with the intent of the playbook author. Sharded jobs are always fine to be simultaneous, because the hosts they operate on have absolutely no overlap. Whether or not a JT can be ran is parallel is (and has always been) up to the playbook author.

File: awx/main/models/workflow.py (hunk 1 above)
Comment (AlanCoding) [reply awx/main/models/workflow.py:466]: we will need a view for the JTs `sharded_jobs` and linked in JT related

**Comment by AlanCoding** [comment]
Hit this bug

```
awx_1        | 13:25:22 celeryd.1   | 2018-08-27 13:25:22,366 ERROR    awx.main.tasks Task awx.main.scheduler.tasks.run_task_manager encountered exception.
awx_1        | 13:25:22 celeryd.1   | Traceback (most recent call last):
awx_1        | 13:25:22 celeryd.1   |   File "/venv/awx/lib/python2.7/site-packages/celery/app/trace.py", line 240, in trace_task
awx_1        | 13:25:22 celeryd.1   |     R = retval = fun(*args, **kwargs)
awx_1        | 13:25:22 celeryd.1   |   File "/venv/awx/lib/python2.7/site-packages/celery/app/trace.py", line 438, in __protected_call__
awx_1        | 13:25:22 celeryd.1   |     return self.run(*args, **kwargs)
awx_1        | 13:25:22 celeryd.1   |   File "/awx_devel/awx/main/scheduler/tasks.py", line 31, in run_task_manager
awx_1        | 13:25:22 celeryd.1   |     TaskManager().schedule()
awx_1        | 13:25:22 celeryd.1   |   File "/awx_devel/awx/main/scheduler/task_manager.py", line 693, in schedule
awx_1        | 13:25:22 celeryd.1   |     finished_wfjs = self._schedule()
awx_1        | 13:25:22 celeryd.1   |   File "/awx_devel/awx/main/scheduler/task_manager.py", line 678, in _schedule
awx_1        | 13:25:22 celeryd.1   |     self.spawn_workflow_graph_jobs(running_workflow_tasks)
awx_1        | 13:25:22 celeryd.1   |   File "/awx_devel/awx/main/scheduler/task_manager.py", line 197, in spawn_workflow_graph_jobs
awx_1        | 13:25:22 celeryd.1   |     job.name = "{} - {}".format(job.name, spawn_node.ancestor_artifacts['job_shard'] + 1)
awx_1        | 13:25:22 celeryd.1   | UnicodeEncodeError: 'ascii' codec can't encode character u'\ud007' in position 27: ordinal not in range(128)
```

**Comment by AlanCoding** [comment]
confirmed that bug will be resolved by

```diff
diff --git a/awx/main/scheduler/task_manager.py b/awx/main/scheduler/task_manager.py
index 4bb0c03d70..c1f71f18d1 100644
--- a/awx/main/scheduler/task_manager.py
+++ b/awx/main/scheduler/task_manager.py
@@ -194,7 +194,7 @@ class TaskManager():
                 kv = spawn_node.get_job_kwargs()
                 job = spawn_node.unified_job_template.create_unified_job(**kv)
                 if 'job_shard' in spawn_node.ancestor_artifacts:
-                    job.name = "{} - {}".format(job.name, spawn_node.ancestor_artifacts['job_shard'] + 1)
+                    job.name = six.text_type("{} - {}").format(job.name, spawn_node.ancestor_artifacts['job_shard'] + 1)
                     job.save()
                 spawn_node.job = job
                 spawn_node.save()
```

File: awx/api/views.py [hunk 4]
```diff
@@ -3142,7 +3142,10 @@ def post(self, request, *args, **kwargs):
             data = OrderedDict()
             data['job'] = new_job.id
             data['ignored_fields'] = self.sanitize_for_response(ignored_fields)
-            data.update(JobSerializer(new_job, context=self.get_serializer_context()).to_representation(new_job))
+            if isinstance(new_job, WorkflowJob):
+                data.update(WorkflowJobSerializer(new_job, context=self.get_serializer_context()).to_representation(new_job))
```
Comment (AlanCoding) [inline awx/api/views.py]: In this case, I can see the `job` field confusing people. I might prefer to make that `None`, or else, if the jobs have already been spawned return a field `jobs` that lists them. I don't think we need to include a literal message explaining what happened, I think the current data structure could act as a red herring for people writing clients.

File: awx/ui/client/features/output/details.component.js [hunk 5]
```diff
@@ -126,6 +126,26 @@ function getSourceWorkflowJobDetails () {
...
+    }
+
+    const shardDetails = resource.model.get('summary_fields.internal_limit.shard');
+
+    if (!shardDetails) {
+        return null;
+    }
+
+    const label = strings.get('labels.SHARD_DETAILS');
+    const offset = `${shardDetails.offset} of ${shardDetails.step} shards`;
```
Comment (one-t) [inline awx/ui/client/features/output/details.component.js]: This is currently using a zero-based value for the offset:
image

File: awx/api/serializers.py (hunk 3 above)
Comment (wenottingham) [reply awx/api/serializers.py]: split, plz.

File: awx/api/serializers.py [hunk 6]
```diff
@@ -3199,6 +3197,13 @@ def to_representation(self, obj):
 
     def get_summary_fields(self, obj):
         summary_fields = super(JobSerializer, self).get_summary_fields(obj)
+        if obj.internal_limit:
+            summary_fields['internal_limit'] = {}
+            if obj.internal_limit.startswith('shard'):
```
Comment (wenottingham) [inline awx/api/serializers.py]: same here, etc. Would like to keep 'shard' out of the UX (including API endpoints).

File: awx/ui/client/features/output/details.component.js (hunk 5 above)
Comment (AlanCoding) [reply awx/ui/client/features/output/details.component.js]: Yes that is correct. It's trivial to +1 to the value, but as of now, I'm unsure if I want to do that on the API side or the UI side. Right now I'm favoring the UI side because these are the parameters we take in the inventory script endpoint. For these kind of aesthetic things, I would like to audit all of the name changes and implement them all at once. The tower-qa tests will also need to be changed at the point.

File: awx/ui/client/src/templates/job_templates/job-template.form.js [hunk 7]
```diff
@@ -271,6 +271,20 @@ function(NotificationsList, i18n) {
                     },
                     ngDisabled: '!(job_template_obj.summary_fields.user_capabilities.edit || canAddJobTemplate)'
                 },
+                job_shard_count: {
+                    label: i18n._('Number of job shards to use'),
```
Comment (jakemcdermott) [inline awx/ui/client/src/templates/job_templates/job-template.form.js]: This might have already been mentioned elsewhere in the review, but we don't want the word 'shard' visible to anyone. The current alternative being used is 'split'.

File: awx/ui/client/features/output/details.component.js (hunk 5 above)
Comment (jakemcdermott) [reply awx/ui/client/features/output/details.component.js]: @one-t just wanted to loop you in - alan and I talked this one out. Current plan is to center it to `1` as you've suggested.

File: awx/api/serializers.py (hunk 3 above)
Comment (dsesami) [reply awx/api/serializers.py]: +1

File: awx/api/serializers.py [hunk 8]
```diff
@@ -3028,6 +3025,7 @@ def get_related(self, obj):
             labels = self.reverse('api:job_template_label_list', kwargs={'pk': obj.pk}),
             object_roles = self.reverse('api:job_template_object_roles_list', kwargs={'pk': obj.pk}),
             instance_groups = self.reverse('api:job_template_instance_groups_list', kwargs={'pk': obj.pk}),
+            sharded_jobs = self.reverse('api:job_template_sharded_jobs_list', kwargs={'pk': obj.pk}),
```
Comment (dsesami) [inline awx/api/serializers.py]: split here, too

File: awx/api/templates/api/inventory_script_view.md [hunk 9]
```diff
@@ -26,6 +26,9 @@ string of `?all=1` to return all hosts, including disabled ones.
 Specify a query string of `?towervars=1` to add variables
 to the hostvars of each host that specifies its enabled state and database ID.
 
+Specify a query string of `?subset=shard2of5` to product an inventory that
```
Comment (dsesami) [inline awx/api/templates/api/inventory_script_view.md]: produce* ?

**Review by dsesami — DISMISSED** [review]

File: awx/api/serializers.py (hunk 3 above)
Comment (AlanCoding) [reply awx/api/serializers.py]: I'm still gathering the totality of renames before implementing them, then going through the process of migrating and integration tests. This is the number of jobs to split into, so the field will probably wind up being `job_split_count`, because I think the UI has been going in that direction.

Also, I think the field pointing from the workflow job to the job template would be best named `split_job_template` as well. This will help keep the naming schemes consistent.

**Comment by matburt** [comment]
recheck

**Comment by matburt** [comment]
recheck

**Comment by AlanCoding** [comment]
@wenottingham @kialam I'm staging the changes for the rename of shard->split. I would like to cover these in a single commit (will update the QE tests at the same time), and IMO it will make the most sense to update the UI at the same time. I hope that I can cover all manual work, then do a fairly automated rename of the rest, then verify both tests passing & UI functionality efficiently.

| Field or text                                 | new value                                             |
|-----------------------------------------------|-------------------------------------------------------|
| `job_template` plus help_text                 | `split_job_template`                                  |
| `job_shard_count` help_text and minimum value | `job_split_count` or `split_job_count` minimum of 1 ✅   |
| `internal_limit` help_text                    | some help text                                        |
| sharded_jobs related link                     | split_jobs ✅                                            |
| internal limit syntax shard0of3               | split1of3 (note, changing to 1 for first) ✅             |
| UI: Edit the shard job template               | Edit the split job template *splitting job template?? ✅  |
| Shard Template                                | Split Template *Split Job Template *Splitlate ✅         |

There are relatively minor decisions remaining, but I still want to get conclusiveness on those.

**Comment by matburt** [comment]
recheck

**Comment by wenottingham** [comment]
Discussing with others, and the intermediate consensus is that 'split' is not the greatest terminology either.

Current suggestions:
1. Job Slicing -> "job slices" -> "slice count" -> "slice 1 of N"
2. Distributed Jobs -> "job slices" -> "slice count" -> "slice 1 of N"

**Comment by AlanCoding** [comment]
Second attempt at an agreeable rename:

| Field or text                                 | new value                                                | even newer value                                                   |
|-----------------------------------------------|----------------------------------------------------------|--------------------------------------------------------------------|
| `job_template` w/o help_text                  | `split_job_template`                                     | `job_template` with help_text                                      |
| `job_shard_count` help_text and minimum value | `job_split_count` or  `split_job_count` minimum of 1 ✅   | `job_slice_count` minimum of 1                                     |
| `internal_limit` help_text                    | some help text                                           | replaced by `job_slice_count` (WJ & J) `job_slice_number` (J only) |
| sharded_jobs related link                     | split_jobs ✅                                             | slice_workflow_jobs ?? (maybe remove)                                             |
| internal limit syntax shard0of3               | split0of3 ✅ (change to 1 for first) ❌                    | slice1of3                                                          |
| UI: Edit the shard job template               | Edit the split job template ✅ *splitting job template??  | Edit the slice job template                                        |
| Shard Template                                | Split Template *Split Job Template *Splitlate ✅          |  Slice Job Template      |                                                            

Scheme for the job & workflow job field names:

```json
# job serialization
{
    "id": 11,
    "type": "job",
    "url": "/api/v2/jobs/11/",
    "related": {...},
    "summary_fields": {...},
    ...
    "job_slice_number": 2,
    "job_slice_count": 5
},
# workflow job serialization
{
    "id": 11,
    "type": "workflow_job",
    "url": "/api/v2/workflow_jobs/11/",
    "related": {...},
    "summary_fields": {...},
    ...
    "job_slice_count": 5
}
```

The related link (previously `sharded_jobs`, then `split_jobs`) would presumably have existed for the UI to use in parallel to the job template RECENT JOBS tab. That's still obtainable through a query filter, and the UI doesn't have any plans of adding such a thing to my knowledge. Maybe we just delete this.

File: awx/main/models/inventory.py [hunk 10]
```diff
@@ -218,67 +221,87 @@ def get_group_children_map(self):
...
                 group_info['vars'] = group.variables_dict
                 data[group.name] = group_info
 
+            # Add ungrouped hosts to all group
+            all_group['hosts'] = [host.name for host in hosts if host.name not in grouped_hosts]
+
+        # Remove any empty groups
+        for group_name in list(data.keys()):
+            if not data.get(group_name, {}).get('hosts', []):
+                data.pop(group_name)
```
Comment (AlanCoding) [inline awx/main/models/inventory.py:298]: This might need to exclude the "all" group, I'm wondering if that's related to a bug @one-t discovered.

**Comment by wenottingham** [comment]
> "replaced by job_slice_count (WJ & J) job_slice_number (J only)"

I thought we were not having the count on the individual slices?

**Comment by wenottingham** [comment]
What is the field/naming for how it is split - does the template define the number of slices, or the size of any one slice?

**Comment by AlanCoding** [comment]
oh, right, I was forgetting about the confusion with non-sliced workflow jobs having an extra integer which would be confusingly 1. The other option for workflow jobs we talked about was to have a boolean field `is_sliced_job`. That's fine with me, but increasingly I am wanting the count to be on the job record as well.

**Comment by one-t** [comment]
I've identified a bug where hosts that are in groups in an inventory will be targeted in every s(lice|plit|hard).

Steps to reproduce:
1. Create an inventory and add some hosts (I did 50)
2. Create an inventory group and add a few to it
3. Run a job template with a split setting > 1
4. Observe in the job details that all group hosts are targeted in every shard
   
I figured this out because I was doing some exploratory testing with user-supplied limits- random hosts were failing when the job was sharded and I used this playbook to identify them

```yaml
---
- hosts: all
  gather_facts: false
  tasks:
    - fail:
      when: ansible_connection != 'local'
```

File: awx/main/models/inventory.py [hunk 11]
```diff
@@ -218,67 +221,87 @@ def get_group_children_map(self):
...
 
             # Build in-memory mapping of groups and their hosts.
-            group_hosts_kw = dict(group__inventory_id=self.id, host__inventory_id=self.id)
-            if 'enabled' in hosts_q:
-                group_hosts_kw['host__enabled'] = hosts_q['enabled']
-            group_hosts_qs = Group.hosts.through.objects.filter(**group_hosts_kw)
-            group_hosts_qs = group_hosts_qs.values_list('group_id', 'host_id', 'host__name')
+            group_hosts_qs = Group.hosts.through.objects.filter(
+                group__inventory_id=self.id,
+                host__inventory_id=self.id
```
Comment (AlanCoding) [inline awx/main/models/inventory.py:264]: here is the source of the bug @one-t mentioned. Group-host relationships are being fetched, even for hosts outside of the slice. That would be fine, if we maintained logic for cutting out those not used when populating the group host lists.

I thought about this logic a lot, because I have another issue in mind where groups aren't populated for smart inventories. The solution to that could be to perform this filter, only restricting to where the host is in the already-fetched host set. However, that would require enumerating the host ids in the query, which is prohibitively expensive. Same applies to prefetching. The same sort of dilemma faces us for the solution of this bug.

**Comment by AlanCoding** [comment]
Update: I will keep to the plan of using the `is_sliced_job` field. This means that orphaned sliced workflow jobs cannot be relaunched. This seems sensible after I thought about it.

File: awx/main/tests/unit/test_tasks.py [hunk 12]
```diff
@@ -246,6 +246,8 @@ def status_side_effect(pk, **kwargs):
             # If `Job.update_model` is called, we're not actually persisting
             # to the database; just update the status, which is usually
             # the update we care about for testing purposes
+            if kwargs.get('result_traceback'):
+                raise Exception('Task encountered error:\n{}'.format(kwargs['result_traceback']))
```
Comment (AlanCoding) [inline awx/main/tests/unit/test_tasks.py:250]: @ryanpetrello I hate to put in drive-by fixes/enhancements like this, but within the course of doing these features, I've hit this a solid handful of times.

### before

```
_______________________________________________________________ TestInventoryUpdateCredentials.test_tower_source[True] _______________________________________________________________
Traceback (most recent call last):
  File "/awx_devel/awx/main/tests/unit/test_tasks.py", line 2216, in test_tower_source
    self.task.run(self.pk)
  File "/awx_devel/awx/main/tasks.py", line 601, in _wrapped
    return f(self, *args, **kwargs)
  File "/awx_devel/awx/main/tasks.py", line 1054, in run
    raise AwxTaskError.TaskError(instance, rc)
Exception: inventory_update None (error) encountered an error (rc=None), please see task stdout for details.
```

### after

```
_______________________________________________________________ TestInventoryUpdateCredentials.test_tower_source[True] _______________________________________________________________
Traceback (most recent call last):
  File "/awx_devel/awx/main/tests/unit/test_tasks.py", line 2218, in test_tower_source
    self.task.run(self.pk)
  File "/awx_devel/awx/main/tasks.py", line 601, in _wrapped
    return f(self, *args, **kwargs)
  File "/awx_devel/awx/main/tasks.py", line 1045, in run
    **extra_update_fields)
  File "/venv/awx/lib/python2.7/site-packages/mock/mock.py", line 1062, in __call__
    return _mock_self._mock_call(*args, **kwargs)
  File "/venv/awx/lib/python2.7/site-packages/mock/mock.py", line 1128, in _mock_call
    ret_val = effect(*args, **kwargs)
  File "/awx_devel/awx/main/tests/unit/test_tasks.py", line 250, in status_side_effect
    raise Exception('Task encountered error:\n{}'.format(kwargs['result_traceback']))
Exception: Task encountered error:
Traceback (most recent call last):
  File "/awx_devel/awx/main/tasks.py", line 944, in run
    env = self.build_env(instance, **kwargs)
  File "/awx_devel/awx/main/tasks.py", line 2054, in build_env
    env['TOWER_LICENSE_TYPE'] = get_licenser().validate()['license_type']
KeyError: 'license_type'
```

**Comment by AlanCoding** [comment]
@kialam It looks like we need to figure out how to integrate these commits into here.

https://github.com/matburt/awx/pull/3

The things still on my agenda with relevance to this branch:

- UI fix for links to workflow jobs in recent jobs
- consider job shard numbers in task impact, unit test
- fix bug with hosts returned in group not in slice

File: awx/main/models/jobs.py [hunk 13]
```diff
@@ -545,6 +589,14 @@ def ansible_virtualenv_path(self):
     def event_class(self):
         return JobEvent
 
+    def copy_unified_job(self, **new_prompts):
+        # Needed for job slice relaunch consistency, do no re-spawn workflow job
+        # target same slice as original job
+        new_prompts['_prevent_slicing'] = True
+        new_prompts.setdefault('_eager_fields', {})
+        new_prompts['_eager_fields']['job_slice_number'] = self.job_slice_number
```
Comment (AlanCoding) [inline awx/main/models/jobs.py:597]: add to this `job_slice_count`, related to https://github.com/ansible/awx/issues/2464 bug

will affect case of launching slice JT, changing the slice count, and then relaunching an old slice. We need to err on the side of the prior job values.

This may also resolve #2464 automatically ... but dunno how I feel about that.

File: awx/api/serializers.py (hunk 8 above)
Comment (AlanCoding) [reply awx/api/serializers.py]: @dsesami could you update your review? the renaming to "slice" is now completed.

File: awx/main/migrations/0050_v340_job_slicing.py [hunk 14]
```diff
@@ -0,0 +1,47 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.11 on 2018-10-15 16:21
+from __future__ import unicode_literals
+
+import awx.main.utils.polymorphic
+from django.db import migrations, models
+import django.db.models.deletion
+
+
+class Migration(migrations.Migration):
```
Comment (ryanpetrello) [inline awx/main/migrations/0050_v340_job_slicing.py]: FYI https://github.com/ansible/awx/pull/2494/files#diff-a8760b06d0784503c5d4b62bd4b95a4b

**Comment by shanemcd** [comment]
recheck

**Review by one-t — APPROVED** [review]
:shipit:

File: awx/ui/client/src/templates/job_templates/job-template.form.js [hunk 15]
```diff
@@ -257,6 +257,19 @@ function(NotificationsList, i18n) {
...
+                    label: i18n._('Job Slicing'),
+                    type: 'number',
+                    integer: true,
+                    min: 1,
+                    default: 1,
+                    spinner: true,
+                    dataTitle: i18n._('Slice Job Count'),
+                    dataPlacement: 'right',
+                    dataContainer: 'body',
+                    awPopOver: "<p>" + i18n._("The number of jobs to slice into at runtime. Will cause the Job Template to launch a workflow if value is greater than 1.") + "</p>",
```
Comment (AlanCoding) [inline awx/ui/client/src/templates/job_templates/job-template.form.js]: @one-t @wenottingham As a documentation matter, we were discussing that this text might need to say something about the intended use case. The intended use case is that the number of slices is almost equal to, or exactly equal to, the number of instances available to run the job template. Logic is then that the job distributes throughout the cluster, or part of the cluster.

We don't want to say all of that, but some mention of number of available instances would seem to address some of QEs concerns about the fact that it is up to the user, themselves, to size this correctly.

Comment (wenottingham) [reply awx/ui/client/src/templates/job_templates/job-template.form.js]: Suggestion:

'Job slice count'

'The number of job slices to create for this job. This should not be set for jobs where cross-host orchestration is done. For best behavior, this number should be high enough that each slice is smaller than an individual cluster node capacity.'

Comment (AlanCoding) [reply awx/ui/client/src/templates/job_templates/job-template.form.js]: > For best behavior, this number should be high enough that each slice is smaller than an individual cluster node capacity.

For best behavior, this number should correspond to the number of execution instances available to this template.

Comment (AlanCoding) [reply awx/ui/client/src/templates/job_templates/job-template.form.js]: maybe "execution instances" -> "execution nodes"... iffy about that

Comment (wenottingham) [reply awx/ui/client/src/templates/job_templates/job-template.form.js]: Surely having more, smaller jobs (50 slices, even if you have only 5 nodes) should be fine. @matburt ?

Comment (matburt) [reply awx/ui/client/src/templates/job_templates/job-template.form.js]: Why are we even talking about slices relative to number of execution nodes? It's entirely possible that you'd end up with multiple jobs on execution nodes. This isn't the kind of thing we need to get into because it'll end up being different for any individual customer's use case. Bill's statement in the 2nd comment is fine.

Comment (AlanCoding) [reply awx/ui/client/src/templates/job_templates/job-template.form.js]: > The number of job slices to create for this job. This should not be set for jobs where cross-host orchestration is done. For best behavior, this number should be high enough that each slice is smaller than an individual cluster node capacity.

Okay, so this is what we want to go with.

Problem, this kind of drops the context of:

> Will cause the Job Template to launch a workflow if value is greater than 1.

So maybe

> The number of job slices to create for this job. Will cause the Job Template to launch a workflow if value is greater than 1. This should not be set for jobs where cross-host orchestration is done. For best behavior, this number should be high enough that each slice is smaller than an individual cluster node capacity.

This seems reasonable, when compared to other tooltips

screen shot 2018-10-30 at 9 03 04 am

**Comment by AlanCoding** [comment]
A couple of pre-merge checks...

```
bash-4.2$ awx-manage makemigrations
No changes detected
```

Good with migrations.

Tests are looking good.

Will push the new tooltip text and rebase when I have agreement on that.

File: awx/ui/client/src/templates/job_templates/job-template.form.js (hunk 15 above)
Comment (wenottingham) [reply awx/ui/client/src/templates/job_templates/job-template.form.js]: I'd drop the bit about a workflow... that's (essentially) an implementation detail that we are surfacing, and we could decide not to in the future.

**Review by jakemcdermott — APPROVED** [review]

**Comment by AlanCoding** [comment]
notes for docs impact:

 - use current docs in section along with JT & workflows
 - add mention in capacity planning

File: awx/ui/client/features/jobs/jobsList.controller.js [hunk 16]
```diff
@@ -76,6 +76,22 @@ function ListJobsController (
             return { icon, link, value };
         });
 
+    vm.getSliceJobDetails = (job) => {
+        if (!job.job_slice_count) {
+            return null;
```
Comment (jakemcdermott) [inline awx/ui/client/features/jobs/jobsList.controller.js:81]: 👍

File: awx/ui/client/features/output/details.component.js [hunk 17]
```diff
@@ -126,6 +126,33 @@ function getSourceWorkflowJobDetails () {
     return { link, tooltip };
 }
 
+function getSliceJobDetails () {
```
Comment (jakemcdermott) [inline awx/ui/client/features/output/details.component.js:129]: 👍

File: awx/ui/client/lib/components/launchTemplateButton/launchTemplateButton.component.js [hunk 18]
```diff
@@ -41,7 +41,13 @@ function atLaunchTemplateCtrl (
                         selectedJobTemplate
                             .postLaunch({ id: vm.template.id })
                             .then(({ data }) => {
-                                $state.go('output', { id: data.job, type: 'playbook' }, { reload: true });
+                                /* Slice Jobs: Redirect to WF Details page if returned
```
Comment (jakemcdermott) [inline awx/ui/client/lib/components/launchTemplateButton/launchTemplateButton.component.js:44]: 👍

File: awx/ui/test/unit/components/job-details-split-jobs.unit.js [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+            get: (obj) => obj.split('.').reduce((i, o) => i && i[o] || null, mockData),
+            has: jasmine.createSpy('has'),
+            options: jasmine.createSpy('options'),
+        },
+        events: {},
+        ws: {}
+    };
+
+    beforeEach(angular.mock.module('at.features.output', ($provide) => {
+        state = {
```
Comment (jakemcdermott) [inline awx/ui/test/unit/components/job-details-split-jobs.unit.js:40]: ⚡️

**Comment by jakemcdermott** [comment]
recheck

**Comment by jakemcdermott** [comment]
recheck

**Comment by jakemcdermott** [comment]
recheck

**Comment by jakemcdermott** [comment]
recheck
